src/server/hll_family.cc

```diff
@@ -0,0 +1,214 @@
+}
+
+OpResult<int64_t> CountHllsSingle(const OpArgs& op_args, string_view key) {
+  auto& db_slice = op_args.shard->db_slice();
+
+  OpResult<PrimeIterator> it = db_slice.Find(op_args.db_cntx, key, OBJ_STRING);
+  if (it.ok()) {
+    string hll;
+    it.value()->second.GetString(&hll);
+    ConvertToDenseIfNeeded(hll);
+    if (isValidHLL(StringToHllPtr(hll)) != HLL_VALID_DENSE) {
+      return OpStatus::INVALID_VALUE;
+    }
+    return pfcountSingle(StringToHllPtr(hll));
+  } else if (it.status() == OpStatus::WRONG_TYPE) {
+    return it.status();
+  } else {
+    // Non existing keys count as 0.
+    return 0;
+  }
+}
+
+void ReadValues(const OpArgs& op_args, ArgSlice keys, vector<OpResult<string>>& values) {
+  auto& db_slice = op_args.shard->db_slice();
+
+  vector<string> hlls;
+  hlls.reserve(keys.size());
+  for (size_t i = 0; i < keys.size(); ++i) {
+    OpResult<PrimeIterator> it = db_slice.Find(op_args.db_cntx, keys[i], OBJ_STRING);
+    if (it.ok()) {
+      string hll;
+      it.value()->second.GetString(&hll);
+      ConvertToDenseIfNeeded(hll);
+      if (isValidHLL(StringToHllPtr(hll)) != HLL_VALID_DENSE) {
+        values.push_back(OpStatus::INVALID_VALUE);
+      } else {
+        values.push_back(std::move(hll));
+      }
+    } else if (it.status() == OpStatus::WRONG_TYPE) {
+      values.push_back(OpStatus::WRONG_TYPE);
+    }
+  }
+}
+
+OpResult<int64_t> PFCountMulti(CmdArgList args, ConnectionContext* cntx) {
+  vector<OpResult<string>> hlls;
+  hlls.reserve(args.size());
+
+  Mutex mu;  // Guards `hlls`
+
+  auto cb = [&](Transaction* t, EngineShard* shard) {
+    ArgSlice shard_args = t->GetShardArgs(shard->shard_id());
+    lock_guard guard(mu);
+    ReadValues(t->GetOpArgs(shard), shard_args, hlls);
+    return OpStatus::OK;
+  };
+
+  Transaction* trans = cntx->transaction;
+  trans->ScheduleSingleHop(std::move(cb));
+
+  vector<HllBufferPtr> ptrs;
+  ptrs.reserve(hlls.size());
+  for (auto& hll : hlls) {
+    if (!hll.ok()) {
+      return hll.status();
+    }
+    ptrs.push_back(StringToHllPtr(hll.value()));
+  }
+  int64_t pf_count = pfcountMulti(ptrs.data(), ptrs.size());
+  if (pf_count < 0) {
+    return OpStatus::INVALID_VALUE;
+  } else {
+    return pf_count;
+  }
+}
+
+void PFCount(CmdArgList args, ConnectionContext* cntx) {
+  if (args.size() == 1) {
+    string_view key = ArgS(args, 0);
+    auto cb = [&](Transaction* t, EngineShard* shard) {
+      return CountHllsSingle(t->GetOpArgs(shard), key);
+    };
+
+    Transaction* trans = cntx->transaction;
+    OpResult<int64_t> res = trans->ScheduleSingleHopT(std::move(cb));
+    HandleOpValueResult(res, cntx);
+  } else {
+    HandleOpValueResult(PFCountMulti(args, cntx), cntx);
+  }
+}
+
+}  // namespace
+
+void HllFamily::Register(CommandRegistry* registry) {
+  using CI = CommandId;
+
+  *registry << CI{"PFADD", CO::WRITE, -3, 1, 1, 1}.SetHandler(PFAdd)
+            << CI{"PFCOUNT", CO::WRITE, -2, 1, -1, 1}.SetHandler(PFCount);
```

Comment:
Should not be write, should it?

Reply:
It actually is a write:

> Note: as a side effect of calling this function, it is possible that the HyperLogLog is modified, since the last 8 bytes encode the latest computed cardinality for caching purposes. So PFCOUNT is technically a write command.
(source: https://redis.io/commands/pfcount/)